File: frontend/src/services/genetics.ts
import api from "./api";

export type Sire = {
  id: string;
  sire_code: string;
  name: string;
  breed_name?: string | null;
  peak_yield_kg?: number | null;
  days_to_peak?: number | null;
  lactation_length_days?: number | null;
  total_milk_yield_kg?: number | null;
  genetic_merit_score: number;
  rank: number;
};

export type SireRankingResponse = {
  sires: Sire[];
  total_sires_evaluated: number;
  top_sire_code?: string | null;
  top_sire_name?: string | null;
  average_sire_total_yield_kg: number;
};

export type CowGeneticProfile = {
  cow_id: string;
  cow_name: string;
  tag_id: string;
  breed_name?: string | null;
  sire_id?: string | null;
  sire_code?: string | null;
  sire_name?: string | null;
  dam_name?: string | null;
  pedigree_status: "Verified Sire Pedigree" | "Estimated from Breed Baseline";
  pedigree_confidence: "High" | "Medium" | "Low";
  estimated_genetic_potential_l: number;
  actual_avg_daily_yield_l?: number | null;
  genetic_merit_rating: number;
  breeding_insights: string[];
};

export type HerdGeneticsSummary = {
  total_cows: number;
  cows_with_pedigree_count: number;
  average_herd_genetic_score: number;
  top_genetic_sire_lines: { sire_name: string; offspring_count: number }[];
  herd_genetic_distribution: Record<string, number>;
  cow_profiles: CowGeneticProfile[];
};

export async function fetchSireRankings(): Promise<SireRankingResponse> {
  const res = await api.get<SireRankingResponse>("/api/v1/genetics/sires");
  return res.data;
}

export async function fetchCowGeneticProfile(cowId: string): Promise<CowGeneticProfile> {
  const res = await api.get<CowGeneticProfile>(`/api/v1/genetics/cow/${cowId}`);
  return res.data;
}

export async function fetchHerdGeneticsSummary(farmId?: string): Promise<HerdGeneticsSummary> {
  const params: Record<string, string> = {};
  if (farmId) params.farm_id = farmId;
  const res = await api.get<HerdGeneticsSummary>("/api/v1/genetics/herd", { params });
  return res.data;
}

File: backend/app/schemas/genetics.py
from __future__ import annotations

from typing import Optional, Literal, Any
from pydantic import BaseModel, ConfigDict


class SireResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    sire_code: str
    name: str
    breed_name: Optional[str] = None
    peak_yield_kg: Optional[float] = None
    days_to_peak: Optional[int] = None
    lactation_length_days: Optional[int] = None
    total_milk_yield_kg: Optional[float] = None
    genetic_merit_score: float
    rank: int = 1


class SireRankingResponse(BaseModel):
    sires: list[SireResponse]
    total_sires_evaluated: int
    top_sire_code: Optional[str] = None
    top_sire_name: Optional[str] = None
    average_sire_total_yield_kg: float


class CowGeneticProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    cow_id: str
    cow_name: str
    tag_id: str
    breed_name: Optional[str] = None
    sire_id: Optional[str] = None
    sire_code: Optional[str] = None
    sire_name: Optional[str] = None
    dam_name: Optional[str] = None
    pedigree_status: Literal["Verified Sire Pedigree", "Estimated from Breed Baseline"]
    pedigree_confidence: Literal["High", "Medium", "Low"]
    estimated_genetic_potential_l: float
    actual_avg_daily_yield_l: Optional[float] = None
    genetic_merit_rating: float
    breeding_insights: list[str]


class HerdGeneticsSummaryResponse(BaseModel):
    total_cows: int
    cows_with_pedigree_count: int
    average_herd_genetic_score: float
    top_genetic_sire_lines: list[dict[str, Any]]
    herd_genetic_distribution: dict[str, int]
    cow_profiles: list[CowGeneticProfileResponse]
